### DISCUSSION TITLE
stop-lying-about-netstandard-2-support/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#80] [Announcements]
# stop-lying-about-netstandard-2-support/

In this post I have a bit of a rant about Microsoft's NuGet packages lying about supporting .NET Standard 2.0 when they kinda don't

https://andrewlock.net/stop-lying-about-netstandard-2-support/

**Comment by chaojian-zhang** [comment]
Nice article! Didn't understand all of it but encountered something similar related to System.Runtime.CompilerServices.Unsafe when I was targeting .Net Framework 4.8 with https://github.com/MiloszKrajewski/K4os.Compression.LZ4.
The thing I wonder is why do they depend on this library😂

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at February 22 2022, 03:20, in reply to [John DeHope](https://disqus.com/by/JohnDeHope3/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/80#discussioncomment-6062974)</em>

---

<p>I actually disagree with this take - .NET Standard worked really well as a way to transition from .NET Framework to .NET Core, I used it extensively. And it <i>still </i>works well. (I'm ignoring the net461 shim and horribleness here that I agree, was a nightmare 😉)</p><p>It's only because Microsoft are going out of their way to <i>pretend</i> the packages support .NET Standard that it's a problem now IMO</p>

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at February 24 2022, 09:42, in reply to [John Zabroski](https://disqus.com/by/john_zabroski/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/80#discussioncomment-6062977)</em>

---

<p>Don't get me wrong, .NET Standard 1.x <i>was</i>, absolutely, a nightmare, I just feel like .NET Standard 2.0 <i>mostly</i> solved that in practical terms. We still ran into out-of-box issues too, not trying to minimze that.</p><p>But these days, if you're on a .NET Framework 4.8 and any .NET Core 2+ (I haven't used UWP etc so can't talk there) it seems like .NET Standard 2.0 has been working pretty well, for what it was trying to achieve?</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at February 23 2022, 06:08, in reply to [Elerius](https://disqus.com/by/elerius/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/80#discussioncomment-6062989)</em>

---

<blockquote>MS can tell the customer if they've set `SuppressTfmSupportBuildWarnings` to get your build to work, we won't offer any support for issues you've encountered with these libraries</blockquote><p></p><p>Except they don't, they say "you can't use this library with .NET Core 2.1, you must pin to the old version of the libraries". <a href="https://github.com/dotnet/runtime/issues/63744#issuecomment-1012293805" rel="nofollow noopener" title="https://github.com/dotnet/runtime/issues/63744#issuecomment-1012293805">https://github.com/dotnet/runtime/issues/63744#issuecomment-1012293805</a></p><p></p><blockquote>Supporting old target frameworks is obviously a hindrance to agility, so I think actions that encourage people to update are good</blockquote><p></p><p>I agree with you completely. What I don't like is the fact that their package says they support .NET Standard 2.0. You can install it into a .NET Core 2.1 package. Everything on NuGet that uses it _looks_ like it will work with .NET Core 2.1. But it doesn't.</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at February 23 2022, 06:05, in reply to [xavi](https://disqus.com/by/disqus_8vfDQIMSwm/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/80#discussioncomment-6062992)</em>

---

<p>It says compatible with .NET Standard 2.0, which means compatible with .NET Core 2.0+ and higher (anything that supports .NET Standard 2.0)</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at February 23 2022, 06:10, in reply to [Amadeusz Sadowski](https://disqus.com/by/amis92/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/80#discussioncomment-6062997)</em>

---

<p>Correct, if you want to use the package, and want to support .NET Core 2.1, then you need to pin to an older version of the package <i>even though the .NET Standard assets appear to work fine</i>.</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at February 23 2022, 07:22, in reply to [Richard Ponton](https://disqus.com/by/richardponton/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/80#discussioncomment-6063003)</em>

---

<p>Hey Richard,</p><p>I totally understand with dropping .NET Core 2.1 support, it makes complete sense (though I'd argue that probably shouldn't have been done in a patch release 😉).</p><p>This rant is entirely directed at the MS packages, and the fact that it now means that .NET Standard 2.0 suddenly doesn't really <i>mean</i> anything any more. I can't look at a NuGet package on <a href="http://nuget.org" rel="nofollow noopener" title="nuget.org">nuget.org</a> and know whether it's supported in a given version or not. That's just a bad user experience IMO.</p><p>Downgrading to the explicit package references is absolutely a partial fix, but that has its own dangers obviously (who knows if you're depending on something that's only in the v6 packages🤷‍♂️ as a consumer, I've got no way of knowing). And the sheer number of packages affected will make this a non-starter.</p><p>Just to re-iterate, totally cool, with .NET Core 2.1 support being dropped, not cool with the packages looking like it <i>is</i> still supported 😄</p>

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at February 23 2022, 07:57, in reply to [Richard Ponton](https://disqus.com/by/richardponton/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/80#discussioncomment-6063006)</em>

---

<blockquote>It went through our entire CI build and unit test process on .NET Core 2.1 with only the netstandard2.0 target without triggering the issue</blockquote><p></p><p>Really! That's interesting 🤔 I wonder how/why it didn't trigger there. buildTransitive is only supported on NuGet 5.0+ (<a href="https://docs.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files#controlling-dependency-assets" rel="nofollow noopener" title="https://docs.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files#controlling-dependency-assets">according to the docs</a>) so I wonder if it's related to that...</p>

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at February 24 2022, 09:39, in reply to [John Zabroski](https://disqus.com/by/john_zabroski/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/80#discussioncomment-6063012)</em>

---

<p>Yeah, I think that must be the difference. We use the latest .NET SDK even when building 2.1 apps. If they're using an earlier version that doesn't support buildTransitive, that would explain the discrepancy.</p><p>Though the fact it all works on .NET Core 2.1 adds further fuel to the fact that the buildTransitive "auto-fail" is an abomination 😅</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at February 24 2022, 09:47, in reply to [Jan J](https://disqus.com/by/janjone/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/80#discussioncomment-6063016)</em>

---

<p>We asked about it here: <a href="https://github.com/dotnet/runtime/issues/63744" rel="nofollow noopener" title="https://github.com/dotnet/runtime/issues/63744">https://github.com/dotnet/runtime/issues/63744</a>. The response was factually incorrect (the packages DO contain .NET Standard assets, that's the point), but clearly they had no interest in addressing it, hence my shouting into the void instead 😉</p>

[a comment by an automated account is not shown]
